File: android/app/src/main/java/com/compartirarchivosred/app/net/DiscoveryService.kt
package com.compartirarchivosred.app.net

import android.content.Context
import android.net.wifi.WifiManager
import com.compartirarchivosred.app.model.Peer
import org.json.JSONObject
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress
import java.net.InetSocketAddress
import java.util.concurrent.ConcurrentHashMap
import kotlin.concurrent.thread

/** Descubrimiento por UDP broadcast en la LAN (puerto 45454). */
class DiscoveryService(
    private val context: Context,
    private val selfId: String,
    private val selfName: String,
    private val onPeersChanged: (List<Peer>) -> Unit
) {
    @Volatile private var running = false
    private var socket: DatagramSocket? = null
    private var multicastLock: WifiManager.MulticastLock? = null
    private val peers = ConcurrentHashMap<String, Peer>()

    fun start() {
        if (running) return
        running = true

        val wifi = context.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
        multicastLock = wifi.createMulticastLock("compartirarchivosred").apply {
            setReferenceCounted(true)
            acquire()
        }

        val s = DatagramSocket(null).apply {
            reuseAddress = true
            broadcast = true
            bind(InetSocketAddress(Proto.DISCOVERY_PORT))
        }
        socket = s

        thread(isDaemon = true) { listenLoop(s) }
        thread(isDaemon = true) { announceLoop(s) }
        thread(isDaemon = true) { expireLoop() }
    }

    private fun announceLoop(s: DatagramSocket) {
        val data = JSONObject()
            .put("v", Proto.VERSION)
            .put("type", Proto.ANNOUNCE)
            .put("id", selfId)
            .put("name", selfName)
            .put("platform", "android")
            .put("port", Proto.TRANSFER_PORT)
            .toString().toByteArray(Charsets.UTF_8)
        val addr = InetAddress.getByName("255.255.255.255")
        while (running) {
            try { s.send(DatagramPacket(data, data.size, addr, Proto.DISCOVERY_PORT)) } catch (_: Exception) {}
            try { Thread.sleep(2000) } catch (_: Exception) {}
        }
    }

    private fun listenLoop(s: DatagramSocket) {
        val buf = ByteArray(2048)
        while (running) {
            try {
                val p = DatagramPacket(buf, buf.size)
                s.receive(p)
                val o = JSONObject(String(p.data, 0, p.length, Charsets.UTF_8))
                if (o.optString("type") != Proto.ANNOUNCE) continue
                val id = o.optString("id")
                if (id.isEmpty() || id == selfId) continue
                val isNew = !peers.containsKey(id)
                peers[id] = Peer(
                    id = id,
                    name = o.optString("name", "?"),
                    platform = o.optString("platform", "?"),
                    host = p.address?.hostAddress ?: "",
                    port = o.optInt("port", Proto.TRANSFER_PORT),
                    lastSeen = System.currentTimeMillis()
                )
                if (isNew) onPeersChanged(snapshot())
            } catch (_: Exception) {}
        }
    }

    private fun expireLoop() {
        while (running) {
            try { Thread.sleep(3000) } catch (_: Exception) {}
            val now = System.currentTimeMillis()
            val before = peers.size
            peers.entries.removeAll { now - it.value.lastSeen > 10000 }
            if (peers.size != before) onPeersChanged(snapshot())
        }
    }

    private fun snapshot(): List<Peer> = peers.values.sortedBy { it.name }

    fun stop() {
        running = false
        try { socket?.close() } catch (_: Exception) {}
        try { multicastLock?.release() } catch (_: Exception) {}
    }
}

File: windows/src/App/Net/DiscoveryService.cs
using System.Collections.Concurrent;
using System.Net;
using System.Net.Sockets;
using System.Text;

namespace CompartirArchivosRED.Net;

public sealed class Peer
{
    public string Id = "";
    public string Name = "";
    public string Platform = "";
    public IPAddress Address = IPAddress.None;
    public int Port;
    public DateTime LastSeen;

    public string Display => $"{Name}  ·  {Platform}  ·  {Address}";
}

/// <summary>Descubrimiento por UDP broadcast en la LAN (puerto 45454).</summary>
public sealed class DiscoveryService : IDisposable
{
    private readonly string _selfId;
    private readonly string _selfName;
    private UdpClient? _udp;
    private CancellationTokenSource? _cts;
    private readonly ConcurrentDictionary<string, Peer> _peers = new();

    public event Action? PeersChanged;

    public DiscoveryService(string selfId, string selfName)
    {
        _selfId = selfId;
        _selfName = selfName;
    }

    public IReadOnlyList<Peer> Peers =>
        _peers.Values.OrderBy(p => p.Name).ToList();

    public void Start()
    {
        _cts = new CancellationTokenSource();
        _udp = new UdpClient();
        _udp.Client.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, true);
        _udp.Client.Bind(new IPEndPoint(IPAddress.Any, Proto.DiscoveryPort));
        _udp.EnableBroadcast = true;

        var ct = _cts.Token;
        Task.Run(() => ListenLoop(ct));
        Task.Run(() => AnnounceLoop(ct));
        Task.Run(() => ExpireLoop(ct));
    }

    private async Task AnnounceLoop(CancellationToken ct)
    {
        var msg = new Msg
        {
            Type = Proto.ANNOUNCE,
            Id = _selfId,
            Name = _selfName,
            Platform = "windows",
            Port = Proto.TransferPort
        };
        byte[] data = Encoding.UTF8.GetBytes(msg.ToJson());
        var ep = new IPEndPoint(IPAddress.Broadcast, Proto.DiscoveryPort);

        while (!ct.IsCancellationRequested)
        {
            try { _udp!.Send(data, data.Length, ep); } catch { }
            try { await Task.Delay(2000, ct); } catch { break; }
        }
    }

    private async Task ListenLoop(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            try
            {
                var result = await _udp!.ReceiveAsync(ct);
                var m = Msg.FromJson(Encoding.UTF8.GetString(result.Buffer));
                if (m == null || m.Type != Proto.ANNOUNCE || m.Id == null) continue;
                if (m.Id == _selfId) continue;

                bool isNew = !_peers.ContainsKey(m.Id);
                _peers[m.Id] = new Peer
                {
                    Id = m.Id,
                    Name = m.Name ?? "?",
                    Platform = m.Platform ?? "?",
                    Address = result.RemoteEndPoint.Address,
                    Port = m.Port ?? Proto.TransferPort,
                    LastSeen = DateTime.UtcNow
                };
                if (isNew) PeersChanged?.Invoke();
            }
            catch (OperationCanceledException) { break; }
            catch { }
        }
    }

    private async Task ExpireLoop(CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            try { await Task.Delay(3000, ct); } catch { break; }
            var now = DateTime.UtcNow;
            bool changed = false;
            foreach (var kv in _peers)
                if ((now - kv.Value.LastSeen).TotalSeconds > 10)
                    changed |= _peers.TryRemove(kv.Key, out _);
            if (changed) PeersChanged?.Invoke();
        }
    }

    public void Dispose()
    {
        try { _cts?.Cancel(); } catch { }
        try { _udp?.Dispose(); } catch { }
    }
}
